### ISSUE TITLE
interpn returns nan when requesting the value at a point which value is not nan

### ISSUE THREAD

**Issue by RubendeBruin** [scipy/scipy#11381]
When I try linear interpolation on the following grid

```
0.0      1.0

nan      1.0 (*)
```

And I request the lower right point (*) then using interpn returns:

- nan when using linear interpolation
- 1.0 when using nearest

I think linear interpolation should return 1 as well since this is a valid solution. This would mean checking if the value at datapoint is requested instead of trying to interpolate it.

I ran into this issue when re-gridding data using xarray on a grid where the bottom row was all nan.
Everytime I would interpolate the row above the nan row would also fill with nans until all my data is gone.

#### Reproducing code example:

```
import numpy as np
from scipy.interpolate import interpn

x = np.array((0,1),dtype=float)
y = np.array((0,1),dtype=float)

points = (x,y)
values = ((0,1),(np.nan,1))

xi = np.array((1,1))

print(interpn(points,values,xi, method='linear'))
print(interpn(points,values,xi, method='nearest'))
```

#### Scipy/Numpy/Python version information:
```
1.3.1 1.16.4 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)
```

**Comment by heisenbuug** [comment]
@RubendeBruin can you explain this more precisely
>  This would mean checking if the value at datapoint is requested instead of trying to interpolate it

**Comment by RubendeBruin** [comment]
@heisenbuug 

The origin of nans in the output of the interpolation is because the interpolation function can not be constructed due to nans in the input.

So evaluation of the interpolation function at a gridpoint will result in nan.

However, at gridpoints the value is known. So there is no need to evaluate the interpolation function when at gridpoints.

Pseudo code:

```
if xi in points:
    return values[xi]
else:
   f = interpolation_function(points,values)
   return f(xi)
```

Diving into the numpy code:

if I re-write https://github.com/scipy/scipy/blob/1e3b2900ed2758c9ad8bc5be55e84b549154a86b/scipy/interpolate/interpolate.py#L2488 to be:

```
    def _evaluate_linear(self, indices, norm_distances, out_of_bounds):
        # slice for broadcasting over trailing dimensions in self.values
        vslice = (slice(None),) + (None,)*(self.values.ndim - len(indices))

        # find relevant values
        # each i and i+1 represents a edge
        edges = itertools.product(*[[i, i + 1] for i in indices])
        values = 0.
        for edge_indices in edges:
            weight = 1.
            for ei, i, yi in zip(edge_indices, indices, norm_distances):
                weight *= np.where(ei == i, 1 - yi, yi)

            print('value = {}, weight = {}'.format(self.values[edge_indices], weight))

            if weight[vslice]!=0:
                values += np.asarray(self.values[edge_indices]) * weight[vslice]
        return values
```

Then this shows:

```
value = [0.], weight = [0.]
value = [1.], weight = [0.]
value = [nan], weight = [0.]
value = [1.], weight = [1.]     <=== This is the only one that matters
```

So it is the 0*nan  (= nan) that causes the issue.

For the example case with only a single input point this can be solved by replacing:

```
values += np.asarray(self.values[edge_indices]) * weight[vslice]
```

with:

```
if weight[vslice] != 0:
    values += np.asarray(self.values[edge_indices]) * weight[vslice]
```

**Comment by Cody-G** [comment]
I think the reason for this behavior is the same as discussed in an issue I opened a while ago in the Julia interpolations package: https://github.com/JuliaMath/Interpolations.jl/issues/192

TL;DR it's a performance optimization.  It avoids the branching logic that would be required to keep the NaN value out of the computation at grid points.

**Comment by ev-br** [comment]
Precisely what @Cody-G said. 

Consider a 1D analog. Given two points, `(x1, y1)` and `(x2, y2)` a linear interpolation for `x1 <= x <= x2` is

```
y = y1*(x2 - x)/(x2 - x1) + y2*(x - x1)/(x2 - x1)
```

Now suppose `y1` is nan.  For `x = x2`, this becomes `nan * 0 + y2*1`. The OP expectation is that the first term is zero. However with the *not-a-number* IEEE754 semantics, the first term is nan (roughly: we don't know what multiplied by a zero is still we don't know what) and that's what scipy.interpolate sticks to. 

So I'm afraid users who want to preserve the input values with exact equality, need to check that exact equality in user code.
